src/ABC_train_test4.py

```diff
@@ -0,0 +1,199 @@
+      generated_y = gen(gen_input) 
+      inputs_fake = torch.cat((x_batch_cuda,generated_y),dim=1).to(device)
+      disc_fake_pred = disc(inputs_fake)
+
+      gen_loss = criterion(disc_fake_pred,real_labels)
+      epoch_loss_gen.append(gen_loss.item())
+
+      #Update gradients 
+      gen_loss.backward()
+      #Update optimizer 
+      gen_opt.step()
+    
+    discriminatorLoss.append(sum(epoch_loss_disc)/len(epoch_loss_disc))
+    generatorLoss.append(sum(epoch_loss_gen)/len(epoch_loss_gen))
+
+  performanceMetrics.plotTrainingLoss2(discriminatorLoss,generatorLoss,np.linspace(1, n_epochs, n_epochs).astype(int))
+
+#Training ABC-GAN Skip Connection 
+def training_GAN_skip_connection(disc,gen,disc_opt,gen_opt,dataset,batch_size,n_epochs,criterion,prior_model,variance,device): 
+  discriminatorLoss = []
+  generatorLoss = []
+  train_loader = DataLoader(dataset, batch_size=batch_size, shuffle=True)
+
+  for epoch in range(n_epochs):
+    epoch_loss_disc = []
+    epoch_loss_gen = []
+    for x_batch,y_batch in train_loader:
+
+      y_shape = list(y_batch.size()) 
+      curr_batch_size = y_shape[0] 
+      y_batch = torch.reshape(y_batch,(curr_batch_size,1)) 
+
+      #Create the labels  
+      real_labels = torch.ones(curr_batch_size,1).to(device)
+      fake_labels = torch.zeros(curr_batch_size,1).to(device)
+
+      #------------------------
+      #Update the discriminator
+      #------------------------
+      disc_opt.zero_grad() 
+
+      #Get discriminator loss for real data 
+      inputs_real = torch.cat((x_batch,y_batch),dim=1).to(device) 
+      disc_real_pred = disc(inputs_real)
+      disc_real_loss = criterion(disc_real_pred,real_labels)
+
+      #Get discriminator loss for fake data
+      gen_input =  pre_generator(x_batch,prior_model,variance,curr_batch_size,device)
+      generated_y = gen(gen_input)  
+      x_batch_cuda = x_batch.to(device)
+      inputs_fake = torch.cat((x_batch_cuda,generated_y),dim=1).to(device) 
+      disc_fake_pred = disc(inputs_fake) 
+      disc_fake_loss = criterion(disc_fake_pred,fake_labels) 
+
+      #Get the discriminator loss 
+      disc_loss = (disc_fake_loss + disc_real_loss) / 2
+      epoch_loss_disc.append(disc_loss.item())
+
+      # Update gradients
+      disc_loss.backward(retain_graph=True)
+      # Update optimizer
+      disc_opt.step()
+
+      #------------------------
+      #Update the Generator 
+      #------------------------
+      gen_opt.zero_grad() 
+
+      #Generate input to generator using ABC pre-generator 
+      gen_input =  pre_generator(x_batch,prior_model,variance,curr_batch_size,device)
+      generated_y = gen(gen_input) 
+      inputs_fake = torch.cat((x_batch_cuda,generated_y),dim=1).to(device)
+      disc_fake_pred = disc(inputs_fake)
+
+      gen_loss = criterion(disc_fake_pred,real_labels)
+      epoch_loss_gen.append(gen_loss.item())
+
+      #Update gradients 
+      gen_loss.backward()
+      #Update optimizer 
+      gen_opt.step()
```

Comment:
yes, no need to apply clamp the skip weight. it is handled in the model itself.
looks good.